substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaUtilZipSubstitutions.java

```diff
@@ -330,32 +435,48 @@ private static long init(boolean nowrap) {
                 if (ret == ZLib.Z_MEM_ERROR()) {
                     throw new OutOfMemoryError();
                 } else {
-                    throw new InternalError();
+                    throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
                 }
             }
         }
     }
 
-    // Inflater.c-Java_java_util_zip_Inflater_setDictionary(JNIEnv *env, jclass cls, jlong addr,
+    // (JDK 11)src/java.base/share/native/libzip/Inflater.c-static void doSetDictionary(JNIEnv *env, jlong addr, jbyte *buf, jint len)
+    private static void doSetDictionary(long addr, CCharPointer bytes, int len) {
+        z_stream strm = WordFactory.pointer(addr);
+        int res = ZLib.inflateSetDictionary(strm, bytes, len);
+        if (res == ZLib.Z_OK()) {
+            return;
+        }
+        if (res == ZLib.Z_STREAM_ERROR() || res == ZLib.Z_DATA_ERROR()) {
+            throw new IllegalArgumentException(CTypeConversion.toJavaString(strm.msg()));
+        }
+        throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
+    }
+
+    // (JDK 11)src/java.base/share/native/libzip/Inflater.c-Java_java_util_zip_Inflater_setDictionary(JNIEnv *env, jclass cls, jlong addr,
+    //                                          jbyteArray b, jint off, jint len)
     @Substitute
     private static void setDictionary(long addr, byte[] b, int off, int len) {
         try (PinnedObject pinned = PinnedObject.create(b)) {
-            CCharPointer bufPlusOff = pinned.addressOfArrayElement(off);
-            int res = ZLib.inflateSetDictionary(WordFactory.pointer(addr), bufPlusOff, len);
-            if (res == ZLib.Z_OK()) {
-                return;
-            } else if (res == ZLib.Z_STREAM_ERROR() || res == ZLib.Z_DATA_ERROR()) {
-                throw new IllegalArgumentException();
-            } else {
-                throw new InternalError();
-            }
+            CCharPointer bytes = pinned.addressOfArrayElement(off);
+            doSetDictionary(addr, bytes, len);
         }
     }
 
-    // Inflater.c-Java_java_util_zip_Inflater_inflateBytes(JNIEnv *env, jobject this, jlong addr,
+    // (JDK 11)src/java.base/share/native/libzip/Inflater.c-Java_java_util_zip_Inflater_setDictionaryBuffer(JNIEnv *env, jclass cls, jlong addr,
+    //                                          jlong bufferAddr, jint len)
+    @Substitute
+    @TargetElement(onlyWith = JDK11OrLater.class)
+    private static void setDictionaryBuffer(long addr, long bufAddress, int len) {
+        doSetDictionary(addr, WordFactory.pointer(addr), len);
```

Comment:
The C code is
```
    jbyte *buf = jlong_to_ptr(bufferAddr);
    doSetDictionary(env, addr, buf, len);
```
but you seem not to be using `bufAddress`.  Should this line be
```
    doSetDictionary(addr, WordFactory(bufAddress), len)
```
?  (I admit to not knowing enough about the zip API to know.)